include/micm/solver/backward_euler.inl

```diff
@@ -0,0 +1,205 @@
+    std::size_t n_convergence_failures = 0;
+    std::size_t cut_in_half_limit = 4;
+    std::size_t cut_in_tenth_limit = 1;
+    const std::array<double, 5> time_step_reductions{ 0.5, 0.5, 0.5, 0.5, 0.1 };
+
+    auto Yn = state.variables_;
+    auto Yn1 = state.variables_;
+
+    Process::UpdateState(processes, state);
+
+    while(t < time_step) {
+      bool converged = true;
+      std::size_t iterations = 0;
+
+      do {
+        // calculate forcing
+        std::fill(forcing.AsVector().begin(), forcing.AsVector().end(), 0.0);
+        process_set.AddForcingTerms(state.rate_constants_, Yn, forcing);
+
+        // calculate jacobian
+        std::fill(state.jacobian_.AsVector().begin(), state.jacobian_.AsVector().end(), 0.0);
+        process_set.SubtractJacobianTerms(state.rate_constants_, Yn, state.jacobian_);
+
+        // subtract the inverse of the time step from the diagonal
+        // TODO: handle vectorized jacobian matrix
+        for(auto& jac : state.jacobian_.AsVector()) {
+          jac *= -1;
+        }
+        for (std::size_t i_block = 0; i_block < state.jacobian_.Size(); ++i_block)
+        {
+          auto jacobian_vector = std::next(state.jacobian_.AsVector().begin(), i_block * state.jacobian_.FlatBlockSize());
+          for (const auto& i_elem : jacobian_diagonal_elements)
+            jacobian_vector[i_elem] -= 1 / H;
+        }
+
+        // We want to solve this equation for a zero
+        // (y_{n+1} - y_n) / H = f(t_{n+1}, y_{n+1})
+
+        // try to find the root by factoring and solving the linear system
+        linear_solver.Factor(state.jacobian_, state.lower_matrix_, state.upper_matrix_, singular);
+
+        auto yn1_iter = Yn1.begin();
+        auto yn_iter = Yn.begin();
+        auto residual_iter = residual.begin();
+        auto forcing_iter = forcing.begin();
+        // forcing_blk in camchem
+        // residual = (Yn1 - Yn) / H - forcing;
+        for(; yn1_iter != Yn1.end(); ++yn1_iter, ++yn_iter, ++residual_iter, ++forcing_iter) {
+          *residual_iter = (*yn1_iter - *yn_iter) / H - *forcing_iter;
+        }
+
+        // the result of the linear solver will be stored in temp
+        // this represnts the change in the solution
+        linear_solver.Solve(residual, temp, state.lower_matrix_, state.upper_matrix_);
+
+        // solution_blk in camchem
+        // Yn1 = Yn1 + temp;
+        auto temp_iter = temp.begin();
+        yn1_iter = Yn1.begin();
+
+        for(; temp_iter != temp.end(); ++temp_iter, ++yn1_iter) {
+          *yn1_iter += *temp_iter;
+        }
+
+        // always make sure the solution is positive regardless of which iteration we are on
+        // remove negatives, accept this solution and continue
+        yn1_iter = Yn1.begin();
+        yn_iter = Yn.begin();
+        for(; yn1_iter != Yn1.end(); ++yn1_iter, ++yn_iter) {
+          *yn_iter = *yn1_iter = std::max(0.0, *yn1_iter);
+        }
+
+        if (iterations == 0) {
```

Comment:
We need to increase `iterations` first. Otherwise it is never updated.

Reply:
Something like:
```
        ++iterations;
        if (iterations == 1) {
          // we always want to accept the first iteration
          continue;
        }
```

Reply:
Ah, true. I have done that on my machine now